src/core/transport/chttp2_transport.c

```diff
@@ -748,6 +753,21 @@ void grpc_chttp2_complete_closure_step(grpc_exec_ctx *exec_ctx,
   *pclosure = NULL;
 }
 
+static void fail_all_outstanding_ops(
+    grpc_exec_ctx *exec_ctx, grpc_chttp2_transport_global *transport_global,
+    grpc_chttp2_stream_global *stream_global) {
+  grpc_chttp2_complete_closure_step(
+      exec_ctx, &stream_global->send_initial_metadata_finished, 0);
+  grpc_chttp2_complete_closure_step(
+      exec_ctx, &stream_global->send_trailing_metadata_finished, 0);
+  grpc_chttp2_complete_closure_step(exec_ctx,
+                                    &stream_global->send_message_finished, 0);
+  grpc_chttp2_complete_closure_step(
+      exec_ctx, &stream_global->recv_initial_metadata_finished, 0);
+  grpc_chttp2_complete_closure_step(
+      exec_ctx, &stream_global->recv_trailing_metadata_finished, 0);
+}
```

Comment:
I think so. This is built on top of #4451 - and that's what was required
for those tests. I'm going to start merging your fixes into this and see if
I can come up with a grand unified PR that addresses all the lost callback
bugs.

On Mon, Dec 14, 2015 at 2:02 PM Sree Kuchibhotla <notifications@github.com>
wrote:

> In src/core/transport/chttp2_transport.c
> <https://github.com/grpc/grpc/pull/4453#discussion_r47565257>:
>
> > @@ -748,6 +753,21 @@ void grpc_chttp2_complete_closure_step(grpc_exec_ctx *exec_ctx,
> >    *pclosure = NULL;
> >  }
> >
> > +static void fail_all_outstanding_ops(
> > +    grpc_exec_ctx *exec_ctx, grpc_chttp2_transport_global *transport_global,
> > +    grpc_chttp2_stream_global *stream_global) {
> > +  grpc_chttp2_complete_closure_step(
> > +      exec_ctx, &stream_global->send_initial_metadata_finished, 0);
> > +  grpc_chttp2_complete_closure_step(
> > +      exec_ctx, &stream_global->send_trailing_metadata_finished, 0);
> > +  grpc_chttp2_complete_closure_step(exec_ctx,
> > +                                    &stream_global->send_message_finished, 0);
> > +  grpc_chttp2_complete_closure_step(
> > +      exec_ctx, &stream_global->recv_initial_metadata_finished, 0);
> > +  grpc_chttp2_complete_closure_step(
> > +      exec_ctx, &stream_global->recv_trailing_metadata_finished, 0);
> > +}
>
> Should recv_message_ready closure also be queued here ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc/pull/4453/files#r47565257>.
>